src/printf.cpp

```diff
@@ -44,18 +52,77 @@ void PrintableBuffer::escape_hex(bool value)
   escape_hex_ = value;
 }
 
-int PrintableCString::print(char *buf, size_t size, const char *fmt)
+int PrintableCString::print(char *buf,
+                            size_t size,
+                            const char *fmt,
+                            ArgumentType)
 {
   return snprintf(buf, size, fmt, value_);
 }
 
-int PrintableInt::print(char *buf, size_t size, const char *fmt)
+int PrintableInt::print(char *buf,
+                        size_t size,
+                        const char *fmt,
+                        ArgumentType expected_type)
 {
-  return snprintf(buf, size, fmt, value_);
+  // Since the value is internally always stored as a 64-bit integer, a cast is
+  // needed to ensure that the type of the argument passed to snprintf matches
+  // the format specifier.
+  // For example, an int64_t argument may be pushed onto the stack while an int
+  // is stored in a register, in which case "%d" would print the wrong value if
+  // we used value_ without an explicit cast.
+  switch (expected_type)
+  {
+    case ArgumentType::CHAR:
+      return snprintf(buf, size, fmt, static_cast<unsigned char>(value_));
+    case ArgumentType::SHORT:
+      return snprintf(buf, size, fmt, static_cast<unsigned short>(value_));
+    case ArgumentType::INT:
+      return snprintf(buf, size, fmt, static_cast<unsigned int>(value_));
+    case ArgumentType::LONG:
+      return snprintf(buf, size, fmt, static_cast<unsigned long>(value_));
+    case ArgumentType::LONG_LONG:
+      return snprintf(buf, size, fmt, static_cast<unsigned long long>(value_));
+    case ArgumentType::INTMAX_T:
+      return snprintf(buf, size, fmt, static_cast<uintmax_t>(value_));
+    case ArgumentType::SIZE_T:
+      return snprintf(buf, size, fmt, static_cast<size_t>(value_));
+    case ArgumentType::PTRDIFF_T:
+      return snprintf(buf, size, fmt, static_cast<ptrdiff_t>(value_));
+    case ArgumentType::POINTER:
+      return snprintf(buf, size, fmt, reinterpret_cast<void *>(value_));
+    default:
```

Comment:
Updated. I originally had `default` here to avoid "control reaches end of non-void function" warnings, but using `__builtin_unreachable()` gets around that (if a value is passed that's not part of the enum, that's undefined behavior anyway).